README.md

```diff
@@ -0,0 +1,190 @@
+# xtm - a library for cross-thread messaging with event loop intergation
+
+The library provides the following facilities:
+
+# xtm_queue
+
+Opaque struct, that represent unidirectional, single-writer-single-reader queue
+implementation with eventloop integration ability.
+
+## xtm_queue_new
+
+Allocation function, used to create instance of struct xtm_queue and return
+its pointer or NULL in case of error. Accepts the queue size as input, which
+must be a power of two, and flags, defining library behavior.
+
+## xtm_queue_delete
+
+Deallocation function, used to free queue and close its internal fds.
+Returns 0 on success. Otherwise -1 with errno set appropriately (as in
+close(2), since it implies close of fds)
+
+## xtm_queue_notify_consumer
+
+Function for queue consumer notification.
+Returns 0 on success. Otherwise -1 with errno set appropriately (as in
+write(2), since it implies write to internal fd)
+
+## xtm_queue_notify_producer
+Function for queue producer notification.
+Returns 0 on success. Otherwise -1 with errno set appropriately (as in
+write(2), since it implies write to internal fd)
+
+## xtm_queue_probe
+
+Returns 0 if queue has space. Otherwise -1 with errno set to ENOBUFS.
+
+## xtm_queue_count
+
+Function returns current message count in queue.
+
+## xtm_queue_push_fun
+
+Function push function and its argument to the queue. This function does not
+notify the consumer thread, but only push to the queue. To notify the consumer
+thread you must call xtm_queue_notify_consumer. The less often you notify, the
+greater the performance, but the greater the latency.
+Returns 0 if queue has space. Otherwise -1 with errno set to ENOBUFS.
+
+## xtm_queue_consumer_fd
+
+Returns file descriptor, that should be watched by consumer thread to
+become readable. When it became readable, consumer should call one of
+the consumer functions: xtm_queue_pop_ptrs or xtm_queue_invoke_funs_all.
+
+## xtm_queue_producer_fd
+Return file descriptor, that should be watched by producer thread to
+become readable. When it became readable, producer may push to queue,
+being sure that there is a place in it.
+
+## xtm_queue_invoke_funs_all
+
+Function calls all functions contained in the queue.
+Return count of invoked functions on success. Otherwise -(count + 1) with errno
+set appropriately (as in write(2), since it implies write to internal fd).
```

Comment:
You need to define what is 'all functions'. That it calls only those functions which were here at the moment of calling this method. And you need to add an explanation why - to prevent infinite loop and starvation. Also you need to mention, that if the queue size is huge (tens of thousands and more), this function still might cause starvation, because it does not allow to invoke a limited number of functions. Again, since you want to write documentation here, it makes little sense to simply narrate what `write(2)` can return. You need to add more details for those who would want to use it without reading all the source code.

Reply:
Fixed